src/gwt/src/org/rstudio/studio/client/rsconnect/ui/RSConnectDeploy.java

```diff
@@ -1457,15 +1462,35 @@ private void showAppError(String error)
    
    private void updateEnvVarsLabel()
    {
-      if (appEnvVars_.isEmpty())
+      if (selectedEnvVars_.isEmpty())
       {
          envVarsLabel_.setText("");
          envVarsLabel_.setVisible(false);
+
+         envVarsList_.setText("");
+         envVarsList_.setVisible(false);
       }
       else
       {
-         envVarsLabel_.setText(constants_.envVarsPublishMessage(appEnvVars_.size(), appContentType()));
+         envVarsLabel_.setText(constants_.envVarsPublishMessage(selectedEnvVars_.size(), appContentType()));
          envVarsLabel_.setVisible(true);
+
+         if (selectedEnvVars_.size() > 3)
+         {
+            String label =
+               selectedEnvVars_.get(0) + ", " +
+               selectedEnvVars_.get(1) + ", " +
+               selectedEnvVars_.get(2) + ", " +
+               "...";
```

Comment:
nit: could use an ellipsis character `…` instead of three dots:  thus saving a tiny bit of space and probably the human race